File: lib/physics.js
/**
 * Termux-Playwright Node.js Physics & Human Interaction Engine
 * Cubic Bézier curves, Fitts's Law easing, muscle jitter, and Gaussian typing delays.
 * @license MIT
 */

'use strict';

/**
 * Standard normal distribution generator using Box-Muller transform.
 * @param {number} mean
 * @param {number} stdDev
 * @returns {number}
 */
function randomGaussian(mean = 0, stdDev = 1) {
    let u = 0, v = 0;
    while (u === 0) u = Math.random();
    while (v === 0) v = Math.random();
    const z = Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
    return mean + z * stdDev;
}

/**
 * Sleep helper utility.
 * @param {number} ms
 * @returns {Promise<void>}
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

class CubicBezierTrajectory {
    static calculateBezierPoint(p0, p1, p2, p3, t) {
        const u = 1.0 - t;
        const tt = t * t;
        const uu = u * u;
        const uuu = uu * u;
        const ttt = tt * t;

        const x = uuu * p0.x + 3 * uu * t * p1.x + 3 * u * tt * p2.x + ttt * p3.x;
        const y = uuu * p0.y + 3 * uu * t * p1.y + 3 * u * tt * p2.y + ttt * p3.y;
        return { x, y };
    }

    static fittsEasing(t) {
        return 0.5 * (1.0 - Math.cos(t * Math.PI));
    }

    /**
     * Generates an array of {x, y} coordinate points simulating human hand movement.
     * @param {{x: number, y: number}|number[]} start
     * @param {{x: number, y: number}|number[]} target
     * @param {Object} [options]
     * @param {number} [options.steps=30]
     * @param {boolean} [options.jitter=true]
     * @param {boolean} [options.overshoot=true]
     * @param {number} [options.deviation=0.25]
     * @returns {Array<{x: number, y: number}>}
     */
    static generateTrajectory(start, target, options = {}) {
        const p0 = Array.isArray(start) ? { x: start[0], y: start[1] } : { x: start.x, y: start.y };
        const p3 = Array.isArray(target) ? { x: target[0], y: target[1] } : { x: target.x, y: target.y };

        const steps = options.steps !== undefined ? options.steps : 30;
        const jitter = options.jitter !== false;
        const overshoot = options.overshoot !== false;
        const deviation = options.deviation || 0.25;

        if (steps <= 1) {
            return [p0, p3];
        }

        const dx = p3.x - p0.x;
        const dy = p3.y - p0.y;
        const dist = Math.hypot(dx, dy);

        const perpX = -dy;
        const perpY = dx;
        const sign1 = Math.random() > 0.5 ? 1 : -1;
        const sign2 = Math.random() > 0.5 ? 1 : -1;

        const dev1 = (Math.random() * (deviation - 0.1) + 0.1) * sign1;
        const dev2 = (Math.random() * (deviation - 0.1) + 0.1) * sign2;

        const p1 = {
            x: p0.x + dx * 0.25 + perpX * dev1,
            y: p0.y + dy * 0.25 + perpY * dev1
        };
        const p2 = {
            x: p0.x + dx * 0.75 + perpX * dev2,
            y: p0.y + dy * 0.75 + perpY * dev2
        };

        const trajectory = [];

        if (overshoot && dist > 40) {
            const overshootDist = Math.random() * Math.min(8.0, dist * 0.08 - 2.0) + 2.0;
            const angle = Math.atan2(dy, dx);
            const overshootP = {
                x: p3.x + Math.cos(angle) * overshootDist,
                y: p3.y + Math.sin(angle) * overshootDist
            };

            const mainSteps = Math.max(2, Math.floor(steps * 0.85));
            const correctSteps = steps - mainSteps;

            for (let i = 0; i < mainSteps; i++) {
                const rawT = i / (mainSteps - 1);
                const easedT = this.fittsEasing(rawT);
                const pt = this.calculateBezierPoint(p0, p1, p2, overshootP, easedT);
                if (jitter && i > 0 && i < mainSteps - 1) {
                    pt.x += randomGaussian(0, 0.4);
                    pt.y += randomGaussian(0, 0.4);
                }
                trajectory.push(pt);
            }

            for (let i = 1; i <= correctSteps; i++) {
                const rawT = i / correctSteps;
                const easedT = this.fittsEasing(rawT);
                const cx = overshootP.x + (p3.x - overshootP.x) * easedT;
                const cy = overshootP.y + (p3.y - overshootP.y) * easedT;
                trajectory.push({ x: cx, y: cy });
            }
        } else {
            for (let i = 0; i < steps; i++) {
                const rawT = i / (steps - 1);
                const easedT = this.fittsEasing(rawT);
                const pt = this.calculateBezierPoint(p0, p1, p2, p3, easedT);
                if (jitter && i > 0 && i < steps - 1) {
                    pt.x += randomGaussian(0, 0.4);
                    pt.y += randomGaussian(0, 0.4);
                }
                trajectory.push(pt);
            }
        }

        trajectory[trajectory.length - 1] = { x: p3.x, y: p3.y };
        return trajectory;
    }
}

class HumanMouse {
    /**
     * @param {any} pageOrMouse
     * @param {number} [currentX=0]
     * @param {number} [currentY=0]
     */
    constructor(pageOrMouse, currentX = 0, currentY = 0) {
        this.pageOrMouse = pageOrMouse;
        this.currentX = currentX;
        this.currentY = currentY;
    }

    _getMouse() {
        return this.pageOrMouse && this.pageOrMouse.mouse ? this.pageOrMouse.mouse : this.pageOrMouse;
    }

    /**
     * Moves mouse naturally to destination coordinates.
     * @param {number} x
     * @param {number} y
     * @param {Object} [options]
     * @returns {Promise<void>}
     */
    async moveTo(x, y, options = {}) {
        const steps = options.steps || 30;
        const jitter = options.jitter !== false;
        const overshoot = options.overshoot !== false;
        const minStepDelay = options.minStepDelay || 0.003;
        const maxStepDelay = options.maxStepDelay || 0.012;

        const trajectory = CubicBezierTrajectory.generateTrajectory(
            { x: this.currentX, y: this.currentY },
            { x, y },
            { steps, jitter, overshoot }
        );

        const mouse = this._getMouse();
        for (const pt of trajectory) {
            if (mouse && typeof mouse.move === 'function') {
                await mouse.move(pt.x, pt.y);
            }
            this.currentX = pt.x;
            this.currentY = pt.y;
            const delay = Math.random() * (maxStepDelay - minStepDelay) + minStepDelay;
            if (delay > 0) {
                await sleep(delay * 1000);
            }
        }

        this.currentX = x;
        this.currentY = y;
    }

    /**
     * Generates and executes trajectory while recording all coordinates.
     * @param {number} startX
     * @param {number} startY
     * @param {number} targetX
     * @param {number} targetY
     * @param {Object} [options]
     * @returns {Promise<Array<{x: number, y: number}>>}
     */
    async moveAndRecord(startX, startY, targetX, targetY, options = {}) {
        this.currentX = startX;
        this.currentY = startY;
        const trajectory = CubicBezierTrajectory.generateTrajectory(
            { x: startX, y: startY },
            { x: targetX, y: targetY },
            options
        );
        const mouse = this._getMouse();
        for (const pt of trajectory) {
            if (mouse && typeof mouse.move === 'function') {
                await mouse.move(pt.x, pt.y);
            }
        }
        this.currentX = targetX;
        this.currentY = targetY;
        return trajectory;
    }

    /**
     * Moves to target coordinates or selector and clicks like a human.
     * @param {string|{x: number, y: number}|number[]} target
     * @param {Object} [options]
     * @returns {Promise<void>}
     */
    async click(target, options = {}) {
        const mouse = this._getMouse();
        let targetX = 0;
        let targetY = 0;

        if (typeof target === 'string') {
            if (this.pageOrMouse && typeof this.pageOrMouse.locator === 'function') {
                const loc = this.pageOrMouse.locator(target).first();
                const box = await loc.boundingBox();
                if (box) {
                    const offX = Math.random() * (box.width * 0.5) + box.width * 0.25;
                    const offY = Math.random() * (box.height * 0.5) + box.height * 0.25;
                    targetX = box.x + offX;
                    targetY = box.y + offY;
                }
            }
        } else if (Array.isArray(target)) {
            targetX = target[0];
            targetY = target[1];
        } else if (target && typeof target.x === 'number') {
            targetX = target.x;
            targetY = target.y;
        }

        await this.moveTo(targetX, targetY, options);

        await sleep(Math.random() * 40 + 20);
        if (mouse && typeof mouse.down === 'function') {
            await mouse.down();
        }
        const holdMs = Math.random() * 100 + 60;
        await sleep(holdMs);
        if (mouse && typeof mouse.up === 'function') {
            await mouse.up();
        }
    }
}

class HumanKeyboard {
    static getGaussianDelay(mean = 0.12, stdDev = 0.035, minD = 0.02, maxD = 0.4) {
        const delay = randomGaussian(mean, stdDev);
        return Math.max(minD, Math.min(maxD, delay));
    }

    /**
     * Types text character by character with human-like Gaussian delays.
     * @param {any} pageOrKeyboard
     * @param {string} text
     * @param {Object} [options]
     * @param {string} [options.selector]
     * @param {number} [options.meanDelay=0.12]
     * @param {number} [options.stdDev=0.035]
     * @param {number} [options.minDelay=0.02]
     * @param {number} [options.maxDelay=0.4]
     * @returns {Promise<void>}
     */
    static async typeText(pageOrKeyboard, text, options = {}) {
        const {
            selector = null,
            meanDelay = 0.12,
            stdDev = 0.035,
            minDelay = 0.02,
            maxDelay = 0.4
        } = options;

        if (selector && pageOrKeyboard && typeof pageOrKeyboard.locator === 'function') {
            await pageOrKeyboard.locator(selector).first().focus();
        }

        const keyboard = pageOrKeyboard && pageOrKeyboard.keyboard ? pageOrKeyboard.keyboard : pageOrKeyboard;

        for (const char of text) {
            if (keyboard && typeof keyboard.type === 'function') {
                await keyboard.type(char);
            }
            let delay = this.getGaussianDelay(meanDelay, stdDev, minDelay, maxDelay);
            if (' ,.!?\n'.includes(char)) {
                delay += Math.random() * 0.1 + 0.05;
            }
            await sleep(delay * 1000);
        }
    }
}

module.exports = {
    CubicBezierTrajectory,
    HumanMouse,
    HumanKeyboard,
    randomGaussian
};

File: termux_playwright/physics.py
"""Physics and human interaction engine for termux-playwright.

Implements Cubic Bézier curve trajectories, Fitts's Law acceleration/deceleration,
micro-jitter, target overshoot, and Gaussian randomized keyboard typing.
All behaviors are parameterized and toggleable.
"""

import asyncio
import math
import random
import time
from dataclasses import dataclass
from typing import List, Tuple, Optional, Union, Any


@dataclass
class Point:
    """Represents a 2D coordinate point."""
    x: float
    y: float


class CubicBezierTrajectory:
    """Cubic Bézier trajectory generator with Fitts's Law easing, overshoot, and jitter."""

    @staticmethod
    def calculate_bezier_point(p0: Point, p1: Point, p2: Point, p3: Point, t: float) -> Point:
        """Calculates a single coordinate point along a cubic Bézier curve at time t in [0, 1]."""
        u = 1.0 - t
        tt = t * t
        uu = u * u
        uuu = uu * u
        ttt = tt * t

        x = uuu * p0.x + 3 * uu * t * p1.x + 3 * u * tt * p2.x + ttt * p3.x
        y = uuu * p0.y + 3 * uu * t * p1.y + 3 * u * tt * p2.y + ttt * p3.y
        return Point(x, y)

    @staticmethod
    def fitts_easing(t: float) -> float:
        """Applies Fitts's Law inspired S-curve easing (acceleration -> steady -> deceleration)."""
        # Sine-based smoothstep S-curve
        return 0.5 * (1.0 - math.cos(t * math.pi))

    @classmethod
    def generate_trajectory(
        cls,
        start: Union[Point, Tuple[float, float]],
        target: Union[Point, Tuple[float, float]],
        steps: int = 30,
        jitter: bool = True,
        overshoot: bool = True,
        deviation: float = 0.25,
    ) -> List[Point]:
        """Generates a list of Point coordinates simulating a natural human mouse stroke.

        Args:
            start: Starting coordinates (Point or tuple of (x, y)).
            target: Destination coordinates (Point or tuple of (x, y)).
            steps: Number of intermediate discrete steps (default 30).
            jitter: Whether to add subtle sub-pixel muscle jitter (default True).
            overshoot: Whether to simulate human target overshooting and correction (default True).
            deviation: Random curve deviation factor for control points (default 0.25).

        Returns:
            List of Point objects forming a non-linear human-like trajectory.
        """
        p0 = Point(start[0], start[1]) if isinstance(start, (tuple, list)) else Point(start.x, start.y)
        p3 = Point(target[0], target[1]) if isinstance(target, (tuple, list)) else Point(target.x, target.y)

        if steps <= 1:
            return [p0, p3]

        dx = p3.x - p0.x
        dy = p3.y - p0.y
        dist = math.hypot(dx, dy)

        # Generate control points with random perpendicular deviation
        perp_x = -dy
        perp_y = dx
        rand_sign1 = 1 if random.random() > 0.5 else -1
        rand_sign2 = 1 if random.random() > 0.5 else -1

        dev1 = random.uniform(0.1, deviation) * rand_sign1
        dev2 = random.uniform(0.1, deviation) * rand_sign2

        p1 = Point(
            p0.x + dx * 0.25 + perp_x * dev1,
            p0.y + dy * 0.25 + perp_y * dev1,
        )
        p2 = Point(
            p0.x + dx * 0.75 + perp_x * dev2,
            p0.y + dy * 0.75 + perp_y * dev2,
        )

        trajectory: List[Point] = []

        # Handle overshoot target
        target_point = p3
        if overshoot and dist > 40:
            overshoot_dist = random.uniform(2.0, min(8.0, dist * 0.08))
            angle = math.atan2(dy, dx)
            overshoot_p = Point(
                p3.x + math.cos(angle) * overshoot_dist,
                p3.y + math.sin(angle) * overshoot_dist,
            )
            main_steps = max(2, int(steps * 0.85))
            correct_steps = steps - main_steps

            for i in range(main_steps):
                raw_t = i / (main_steps - 1)
                eased_t = cls.fitts_easing(raw_t)
                pt = cls.calculate_bezier_point(p0, p1, p2, overshoot_p, eased_t)
                if jitter and 0 < i < main_steps - 1:
                    pt.x += random.gauss(0, 0.4)
                    pt.y += random.gauss(0, 0.4)
                trajectory.append(pt)

            # Correction stroke to true target
            for i in range(1, correct_steps + 1):
                raw_t = i / correct_steps
                eased_t = cls.fitts_easing(raw_t)
                cx = overshoot_p.x + (p3.x - overshoot_p.x) * eased_t
                cy = overshoot_p.y + (p3.y - overshoot_p.y) * eased_t
                trajectory.append(Point(cx, cy))
        else:
            for i in range(steps):
                raw_t = i / (steps - 1)
                eased_t = cls.fitts_easing(raw_t)
                pt = cls.calculate_bezier_point(p0, p1, p2, target_point, eased_t)
                if jitter and 0 < i < steps - 1:
                    pt.x += random.gauss(0, 0.4)
                    pt.y += random.gauss(0, 0.4)
                trajectory.append(pt)

        # Ensure exact end point matches destination
        trajectory[-1] = Point(p3.x, p3.y)
        return trajectory


class HumanMouse:
    """Simulates realistic human mouse behavior using Bézier curves and Fitts's Law."""

    def __init__(self, page_or_mouse: Any, current_x: float = 0.0, current_y: float = 0.0):
        self.page_or_mouse = page_or_mouse
        self.current_x = current_x
        self.current_y = current_y

    def _get_mouse(self) -> Any:
        if hasattr(self.page_or_mouse, "mouse"):
            return self.page_or_mouse.mouse
        return self.page_or_mouse

    async def move_to(
        self,
        x: float,
        y: float,
        steps: int = 30,
        jitter: bool = True,
        overshoot: bool = True,
        min_step_delay: float = 0.003,
        max_step_delay: float = 0.012,
    ) -> None:
        """Moves the mouse naturally to (x, y) asynchronously."""
        mouse = self._get_mouse()
        trajectory = CubicBezierTrajectory.generate_trajectory(
            start=Point(self.current_x, self.current_y),
            target=Point(x, y),
            steps=steps,
            jitter=jitter,
            overshoot=overshoot,
        )

        for pt in trajectory:
            if hasattr(mouse, "move"):
                await mouse.move(pt.x, pt.y)
            self.current_x = pt.x
            self.current_y = pt.y
            delay = random.uniform(min_step_delay, max_step_delay)
            if delay > 0:
                await asyncio.sleep(delay)

        self.current_x = x
        self.current_y = y

    def move_to_sync(
        self,
        x: float,
        y: float,
        steps: int = 30,
        jitter: bool = True,
        overshoot: bool = True,
        min_step_delay: float = 0.003,
        max_step_delay: float = 0.012,
    ) -> None:
        """Moves the mouse naturally to (x, y) synchronously."""
        mouse = self._get_mouse()
        trajectory = CubicBezierTrajectory.generate_trajectory(
            start=Point(self.current_x, self.current_y),
            target=Point(x, y),
            steps=steps,
            jitter=jitter,
            overshoot=overshoot,
        )

        for pt in trajectory:
            if hasattr(mouse, "move"):
                mouse.move(pt.x, pt.y)
            self.current_x = pt.x
            self.current_y = pt.y
            delay = random.uniform(min_step_delay, max_step_delay)
            if delay > 0:
                time.sleep(delay)

        self.current_x = x
        self.current_y = y

    async def move_and_record(
        self,
        start_x: float,
        start_y: float,
        target_x: float,
        target_y: float,
        steps: int = 30,
        jitter: bool = True,
        overshoot: bool = True,
    ) -> List[Point]:
        """Generates and executes trajectory while recording all intermediate coordinates."""
        self.current_x = start_x
        self.current_y = start_y
        trajectory = CubicBezierTrajectory.generate_trajectory(
            start=Point(start_x, start_y),
            target=Point(target_x, target_y),
            steps=steps,
            jitter=jitter,
            overshoot=overshoot,
        )
        mouse = self._get_mouse()
        for pt in trajectory:
            if hasattr(mouse, "move"):
                await mouse.move(pt.x, pt.y)
        self.current_x = target_x
        self.current_y = target_y
        return trajectory

    async def click(
        self,
        target: Union[str, Tuple[float, float], Point],
        steps: int = 30,
        jitter: bool = True,
        overshoot: bool = True,
        click_delay_min: float = 0.06,
        click_delay_max: float = 0.16,
    ) -> None:
        """Moves to target coordinates/selector and performs a human-like mouse click asynchronously."""
        mouse = self._get_mouse()
        x, y = 0.0, 0.0

        if isinstance(target, str):
            # Selector string
            if hasattr(self.page_or_mouse, "locator"):
                loc = self.page_or_mouse.locator(target).first
                box = await loc.bounding_box()
                if box:
                    # Randomize click point inside element bounds
                    offset_x = random.uniform(box["width"] * 0.25, box["width"] * 0.75)
                    offset_y = random.uniform(box["height"] * 0.25, box["height"] * 0.75)
                    x = box["x"] + offset_x
                    y = box["y"] + offset_y
        elif isinstance(target, (tuple, list)):
            x, y = float(target[0]), float(target[1])
        elif isinstance(target, Point):
            x, y = target.x, target.y

        await self.move_to(x, y, steps=steps, jitter=jitter, overshoot=overshoot)

        # Dwell time before down
        await asyncio.sleep(random.uniform(0.02, 0.06))
        if hasattr(mouse, "down"):
            await mouse.down()
        # Hold duration
        await asyncio.sleep(random.uniform(click_delay_min, click_delay_max))
        if hasattr(mouse, "up"):
            await mouse.up()


class HumanKeyboard:
    """Simulates realistic human typing with Gaussian randomized keystroke delays."""

    @staticmethod
    def get_gaussian_delay(mean: float = 0.12, std_dev: float = 0.035, min_d: float = 0.02, max_d: float = 0.4) -> float:
        """Returns a Gaussian-distributed keystroke delay clamped within [min_d, max_d]."""
        delay = random.gauss(mean, std_dev)
        return max(min_d, min(max_d, delay))

    @classmethod
    async def type_text(
        cls,
        page_or_keyboard: Any,
        text: str,
        selector: Optional[str] = None,
        mean_delay: float = 0.12,
        std_dev: float = 0.035,
        min_delay: float = 0.02,
        max_delay: float = 0.4,
    ) -> None:
        """Types text character by character with human-like Gaussian delays asynchronously."""
        if selector and hasattr(page_or_keyboard, "locator"):
            await page_or_keyboard.locator(selector).first.focus()

        keyboard = page_or_keyboard.keyboard if hasattr(page_or_keyboard, "keyboard") else page_or_keyboard

        for char in text:
            if hasattr(keyboard, "type"):
                await keyboard.type(char)
            delay = cls.get_gaussian_delay(mean_delay, std_dev, min_delay, max_delay)
            # Occasional pause for punctuation or spaces
            if char in " ,.!?\n":
                delay += random.uniform(0.05, 0.15)
            await asyncio.sleep(delay)

    @classmethod
    def type_text_sync(
        cls,
        page_or_keyboard: Any,
        text: str,
        selector: Optional[str] = None,
        mean_delay: float = 0.12,
        std_dev: float = 0.035,
        min_delay: float = 0.02,
        max_delay: float = 0.4,
    ) -> None:
        """Types text character by character with human-like Gaussian delays synchronously."""
        if selector and hasattr(page_or_keyboard, "locator"):
            page_or_keyboard.locator(selector).first.focus()

        keyboard = page_or_keyboard.keyboard if hasattr(page_or_keyboard, "keyboard") else page_or_keyboard

        for char in text:
            if hasattr(keyboard, "type"):
                keyboard.type(char)
            delay = cls.get_gaussian_delay(mean_delay, std_dev, min_delay, max_delay)
            if char in " ,.!?\n":
                delay += random.uniform(0.05, 0.15)
            time.sleep(delay)
